site/app/controllers/admin/LateController.php

```diff
@@ -98,7 +98,24 @@ public function updateLateDays($csv_option = null) {
 
             $date_time = DateUtils::parseDateTime($_POST['datestamp'], $this->core->getUser()->getUsableTimeZone());
 
-            $this->core->getQueries()->updateLateDays($_POST['user_id'], $date_time, $_POST['late_days']);
+            // This is by all means an ugly hack but this same process is completed elsewhere while loading this page
+            // and there is no way to get a specific user by ID and get the number of late days at the same time.
+            // Ultimately, the performance difference is rather negligible since only instructors load this page.
```

Comment:
Why not just add a new query function that gives you the specific user?

Reply:
I could, but that requires modification of the database files.  I do not know SQL well enough to do that in a manner such that I am certain that there are no vulnerabilities or points of failure.  It is also worth noting that all of the users are accessed in multiple other places on this page and we can assume that there are is a conservative maximum of 1000 students being iterated through which is relatively small given that only instructors use this page and they only use it occasionally.

Reply:
You should not worry about messing up the SQL in such a way that would introduce a security failure or point of failure. That will ideally get caught and ironed out in the code review process. I'm happy to give information on the SQL interface or practices or whatever you might need so feel free to ask on Slack. Really, the only big main thing to do is just always use escaped queries (i.e. with the `?` for your input and then pass the variables in as an array to `query()` method).

Anyway, here's a SQL function you can use to accomplish what you want:

```
    public function getLateDaysForUserOnTimestamp(string $user_id, \DateTime $since_timestamp): ?SimpleLateUser {
        $this->course_db->query(
            "
        SELECT u.user_id, user_firstname, user_preferred_firstname,
          user_lastname, user_preferred_lastname, allowed_late_days, since_timestamp
        FROM users AS u
        FULL OUTER JOIN late_days AS l
          ON u.user_id=l.user_id
        WHERE allowed_late_days IS NOT NULL
          AND u.user_id = ?
          AND since_timestamp = ?
        ORDER BY
          user_email ASC, since_timestamp DESC;",
          [$user_id, $since_timestamp)]
        );

        return $this->course_db->getRowCount() === 1
            ? new SimpleLateUser($this->core, $this->course_db->row())
            : null;
    }
```

The modification for this function from `getUsersWithLateDays` is the introduction of the two new WHERE clauses on `u.user_id` and `since_timestamp`, which we use to get a singular row from the DB that fit those criteria. The \PDO adapter handles converting the `\DateTime` object into something that slots into the query, and the postgres will handle timezone conversions on the comparison for us. If no such row exists, then we just return null.

Reply:
Thanks!  I have made the requested changes.